### PR TITLE
(mostly) fix spectator and fix players spawning inside blocks

### PR DESCRIPTION
**Pull request by dany-on-demand**
You might want to skim through the code first. Also I had some formatter installed and it changed some other formatting, but it's fine I hope. I couldn't find any way to break spectator mode now but I don't know all the edge cases in my head. 

Tested in BeS, OS

Fixes:
- spectator mode completely broken
- you can now view players POV too
Changes proposed in this pull request:
- z spawn offset used same as in pyspades, resulting in the familiar "drop on spawn"

### CODE REVIEW

File: Source/Util/Enums.h [hunk 1]
```diff
@@ -81,7 +81,7 @@ typedef enum {
 typedef enum {
     TEAM_A         = 0,
     TEAM_B         = 1,
-    TEAM_SPECTATOR = 2,
+    TEAM_SPECTATOR = 255,
```
Comment (dany-on-demand) [inline Source/Util/Enums.h:84]: Herein lies the fix. I know, it made me very emotional too

File: Source/Server/Player.c [hunk 2]
```diff
@@ -227,7 +227,7 @@ void init_player(server_t*  server,
     player->allow_killing                        = 1;
     player->allow_team_killing                   = 0;
     player->muted                                = 0;
-    player->team                                 = 2;
+    player->team                                 = TEAM_SPECTATOR;
```
Comment (dany-on-demand) [inline Source/Server/Player.c:230]: I think this is different from the pyspades behaviour? Should check

File: Source/Server/Player.c [hunk 3]
```diff
@@ -149,22 +149,22 @@ void set_player_respawn_point(server_t* server, player_t* player)
         player->movement.forward_orientation.z = 0.f;
 
         // Find nearest spawn point
-        for(uint16_t z = player->movement.position.z; z < server->s_map.map.size_z; z++) {
-            if(is_valid_spawn_point(server, player->movement.position.x, player->movement.position.y, z)) {
-                player->movement.position.z = z;
+        for (uint16_t z = player->movement.position.z; z < server->s_map.map.size_z; z++) {
+            if (is_valid_spawn_point(server, player->movement.position.x, player->movement.position.y, z)) {
+                player->movement.position.z = z - 2.4F;
```
Comment (dany-on-demand) [inline Source/Server/Player.c:154]: Ref: https://github.com/piqueserver/piqueserver/blob/b14f6f00ca021ef724cb72e975bd8228a25dcc89/pyspades/player.py#L847C20-L847C20

I'm not sure about this up and down looping?

File: Source/Server/Packets/IntelPickup.c [hunk 4]
```diff
@@ -7,10 +7,10 @@ void send_intel_pickup(server_t* server, player_t* player)
         return;
     }
     uint8_t team;
-    if (player->team == 0) {
-        team = 1;
+    if (player->team == TEAM_A) {
```
Comment (dany-on-demand) [inline Source/Server/Packets/IntelPickup.c:10]: Ugly and copypasted everywhere. Ternary?

File: Source/Server/Packets/KillAction.c [hunk 5]
```diff
@@ -17,6 +17,11 @@ void send_kill_action_packet(server_t* server,
     if (player->alive == 0) {
         return; // Cant kill player if they are dead
     }
+
+    if (player->has_intel) {
+        send_intel_drop(server, player);
```
Comment (dany-on-demand) [inline Source/Server/Packets/KillAction.c:22]: Note: you need to send the intel drop packet BEFORE the kill packet. Pyspades does the same and I don't remember the reasons for this but it makes sense logically.

File: Source/Server/Packets/ChangeTeam.c [hunk 6]
```diff
@@ -4,24 +4,34 @@
...
-    player->state = STATE_WAITING_FOR_RESPAWN;
+
+    if (server->protocol.num_team_users[old_team] > 0)
+        server->protocol.num_team_users[old_team]--;
+
+    player->team = team;
+    server->protocol.num_team_users[team]++;
+
+    if (old_team == TEAM_SPECTATOR) {
+        send_respawn(server, player);
```
Comment (dany-on-demand) [inline Source/Server/Packets/ChangeTeam.c:33]: Instant respawn in this case. Matches behaviour in pyspades

File: Source/Server/Packets/ChangeTeam.c [hunk 7]
```diff
@@ -4,24 +4,34 @@
...
-    player->team = team;
     if (player->id != received_id) {
         LOG_WARNING("Assigned ID: %d doesnt match sent ID: %d in change team packet", player->id, received_id);
     }
-    server->protocol.num_team_users[player->team]++;
-    send_kill_action_packet(server, player, player, 5, 5, 0);
-    player->state = STATE_WAITING_FOR_RESPAWN;
+
+    if (server->protocol.num_team_users[old_team] > 0)
+        server->protocol.num_team_users[old_team]--;
```
Comment (dany-on-demand) [inline Source/Server/Packets/ChangeTeam.c:27]: Previous behaviour resulted in
num_team_users {
TEAM_A: 1
TEAM_B: 2
TEAM_SPECTATOR: 255, 254, 253....
}

File: Source/Server/Packets/ChangeTeam.c [hunk 8]
```diff
@@ -4,24 +4,34 @@
 
 void receive_change_team(server_t* server, player_t* player, stream_t* data)
 {
+    // @Todo: if player is not initialised yet (pyspades uses if not self.name) ignore this
```
Comment (dany-on-demand) [inline Source/Server/Packets/ChangeTeam.c:7]: not 100% sure about this. feel free to leave this out

**Review by Haxk20 — APPROVED** [review]
Looks good. One of the commits should have just Source: instead of listing of the files but that is good enough.

**Pull request merged**